### ISSUE TITLE
Wrong issue in rule typescript:S3516 "Function returns should not be invariant"

### ISSUE THREAD

**Issue by AMorgaut** [SonarSource/eslint-plugin-sonarjs#323]
**I want to report a bug.**

**Reproducer**

```javascript
function updateDetails(details, label, value) {
    const item = details.displayItems.
        find((elem) => elem.label === label);

    if (!item) {
        return details;
    }

    item.value = String(value);
    return details;
}
```

Sonar report a S3516 Typescript issue "Function returns should not be invariant"

**Expected behavior**

The exposed code use the "fail fast" / "exit early" pattern
It returns the same object in both exit points but 
- `details` was not updated when returned at the 1st exit point
- while it got modified before being returnd in the 2d exit point.

**Rule key:** RSPEC-3516 / typescript:S3516

**Comment by vilchik-elena** [comment]
Why don't you do? seems more readable
```js
    if (item) {
      item.value = String(value);
    }
    return details;
```

I still agree it's a FP, but I don't think we can improve the rule to avoid such FPs.

**Comment by Wohops** [comment]
This issue has been migrated to Jira. [ESLINTJS-19](https://sonarsource.atlassian.net/browse/ESLINTJS-19)

**Comment by AMorgaut** [comment]
> Why don't you do? seems more readable
> 
> ```js
>     if (item) {
>       item.value = String(value);
>     }
>     return details;
> ```
> 
> I still agree it's a FP, but I don't think we can improve the rule to avoid such FPs.

Because this code exemple is a simplified code
It use the "fail fast" / "exit fast" pattern allowing to have 
- clear out of complexity exits
- the least possible indented code

In real use cases more code would go there
``` 
    // more code working on the value
    item.value = String(value);
    return details;
```

Some could argue they don't like such code beacause it is not pure function
I admit I would agree
but still the rule is not about pure functions which are unfortunatelly far from being a standard

**Comment by AMorgaut** [comment]
this issue is similar to this python issue
https://sonarsource.atlassian.net/browse/SONARPY-787
